android/app/src/main/java/com/expensify/chat/customairshipextender/CustomNotificationProvider.java

```diff
@@ -220,30 +210,72 @@ private void applyMessageStyle(@NonNull Context context, NotificationCompat.Buil
                         .setGroupConversation(true)
                         .setConversationTitle(conversationName);
 
+
                 // Add all conversation messages to the notification, including the last one we just received.
-                for (NotificationMessage cachedMessage : notificationData.messages) {
-                    messagingStyle.addMessage(cachedMessage.text, cachedMessage.time, notificationData.getPerson(cachedMessage.accountID));
+                NotificationCompat.MessagingStyle previousStyle = NotificationCompat.MessagingStyle.extractMessagingStyleFromNotification(existingReportNotification.getNotification());
+                List<NotificationCompat.MessagingStyle.Message> messages = previousStyle != null ? previousStyle.getMessages() : new ArrayList<>(List.of(recreatePreviousMessage(existingReportNotification)));
```

Comment:
I'd consider using `HashMap` instead, since it has lower complexity of addition operation and we don't take the advantage of better access complexity of `ArrayList`.